vllm/v1/worker/gpu_model_runner.py

```diff
@@ -2018,9 +2018,19 @@ def initialize_kv_cache(self, kv_cache_config: KVCacheConfig) -> None:
                         num_blocks, kv_cache_spec.block_size,
                         kv_cache_spec.num_kv_heads, kv_cache_spec.head_size)
                     dtype = kv_cache_spec.dtype
-                    kv_caches[layer_name] = torch.zeros(kv_cache_shape,
-                                                        dtype=dtype,
-                                                        device=self.device)
+                    # TODO proper impl handling MLA
+                    if len(kv_cache_shape) > 3:
+                        kv_cache_shape = (2, kv_cache_shape[1],
+                                          kv_cache_shape[3], kv_cache_shape[2],
+                                          kv_cache_shape[4])
+                        kv_caches[layer_name] = torch.zeros(
+                            kv_cache_shape, dtype=dtype,
+                            device=self.device).permute(0, 1, 3, 2, 4)
+                    else:
+                        # MLA default layout is fine
+                        kv_caches[layer_name] = torch.zeros(kv_cache_shape,
```

Comment:
this is the bit to refactor with https://github.com/vllm-project/vllm/pull/18775